### ISSUE TITLE
Add  config.General.activity in jobClassAd 

### ISSUE THREAD

**Issue by todor-ivanov** [dmwm/CRABServer#5666]
The config parameter 'conconfig.General.activity' needs to be added to the jobClassAd in order to be visible to MONIT.

**Comment by belforte** [comment]
this is starting to be really needed. See also
https://its.cern.ch/jira/browse/CMSMONIT-28

**Comment by belforte** [comment]
I think we nee to add
```
 ('CRAB_Activity', 'activity'),
```
after this line
https://github.com/dmwm/CRABServer/blob/f3f0acac35568c65cf67cbc576a89da9e6e07407/src/python/TaskWorker/Actions/DagmanSubmitter.py#L44
and
```
info['activity'] = task['activity']
or better yet, since usually there is no activity, set the classAd to the string "None"
(or whatever we use as default also for other classAds to indicate no value)
info['activity'] = task['tm_activity'] or "None"
```
afte e.g. this line
https://github.com/dmwm/CRABServer/blob/f3f0acac35568c65cf67cbc576a89da9e6e07407/src/python/TaskWorker/Actions/DagmanSubmitter.py#L384

**Comment by belforte** [comment]
as to the classAd name, I'd like to use `CMS_JobType` as already used by Production to tell processing from lofharvesting and merge etc. for example
https://monit-grafana.cern.ch/d/000000628/cms-job-monitoring?orgId=11&refresh=15m
but the problem is that ES feeding code massage the classAd when sending things to ES so that for
CRAB is likelty to be overwritten with `Analysis` no matter what.
https://github.com/dmwm/cms-htcondor-es/blob/cff802ae4204e58467dff40f6ae77236a095ed87/src/htcondor_es/convert_to_json.py#L591

This was surely a sensible first shot at it when @bbockelm initially coded it, but what do we do now ?
@vkuznet can you and Federica take over define proper classAd names to use here ?

I'd like to stop defining "my own classAds"  !

**Comment by leggerf** [comment]
I don't have the knowledge of who is using downstream CMS_JobType so I cannot judge implications if you change it. @stiegerb what do you think? Why do you need to rename it?

**Comment by belforte** [comment]
@leggerf it is only used for Benjamin's dashboard so far, but as we build more dashboards its use should grow.

**Comment by stiegerb** [comment]
For me there's no reason not to split this into further subcategories, but probably @bbockelm should comment on the original logic. The question is essentially whether we want to be able to visualize (in the same plot) analysis jobs in one group, and production jobs split into sub groups, or whether we'll just keep everything separate.

**Comment by leggerf** [comment]
ok, I think it would be useful to have two attributes per job: one to distinguish if it is analysis, production, validation, hammercloud? (here you need to help me if there are more types I'm missing..hammercloud for example in ATLAS we have several types). Then another attribute should define what the job executes: reco, sim, evtgen, etc. For this one can also use the input type to distinguish (if this information is saved)

**Comment by belforte** [comment]
@leggerf 
we may ave to cope a bit with what exist, even if a revolution may be appealing.
Starting point: define what CMS_JobType should be.

My 20c follow:
Current values are an odd mix:
`Cleanup,Express,Harvesting,LogCollect,Merge,Processing,Production, Repack,Skim `
Some descrive WMA internals(merge, LogCollect) some different physics use cases (Skim, Processing) some operational areas (Express)
Maybe we can generalike a bit into something like "activity" so that we can add
`hammercloud, useranalyis, userMC` and if we are ambitious pack a bit the
above list so that e.g. `WMA service` contains e.g. merge, cleanup, logcollect... stuff
that does not produce useful output. We may e.g. find useful to split CRAB jobs
in more subcategories as well (overflown, using custom script, using use code or only
config...),, all considerations that pint to keeping the CMS_JobType list small and
fanciers subdivisions should better be made at plot time using filters.
Than if we want to distinguish sim/gen/reco... (and e.g. help us to tell DIGIRECO from GEN)
and classical pileup from premixing) we will need a new attribute.

Definitely we want to have plots with production vs. analyusis w/o hacing all WMA subcategotries, but for that it may be enoug to separate by submission tool, like old dashboard was doing.

**Comment by stiegerb** [comment]
Just to point out: there's nothing stopping us from just adding an `activity` field. We just need to define it in a smart (and robust) way.

Note also that we don't currently have a field for the submission tool.

**Comment by belforte** [comment]
I'd like to hear from whoever defined CMS_JobType (@amaltaro ?) what they wanted it to mean, and maybe we can add 3 or 4 job types for analysis instead of forcing all CRAB submissions into CMS_JobType=analysis. Sort of, forget why Brian did was he did in that one line and have a fresh look. In the we can
always combine and relable things when the plot. Simply a matter of documenting what our ads mean.

**Comment by leggerf** [comment]
Who/what is responsible for ClassAds in general?

**Comment by belforte** [comment]
great question @leggerf 
so far we follow CMS bazaar style. Everybody does what they like (WMA, CRAB, SI) until/ulnless Brian tells them not to.  think SI does not really care unless we add so many to be a problem, and WMA/CRAB should better find a common language, if Stefano+Alan+Valentin agree, I am sure Brian will not mind being left out. All in all now we only want to make sure of CRAB/WMA consistency and proper documentation. If monitoring eventually need something which is not reported, we'll have to implement in both.

**Comment by vkuznet** [comment]
and, since we're talking about particular attribute, where the description of existing one? A particularly good example for HTCondor ones can be found here: https://github.com/dmwm/cms-htcondor-es/blob/master/README.md and I would suggest to create similar one for CRAB ones.

**Comment by amaltaro** [comment]
Looking at the HTCondor attributed documentation, this is what matters for this ticket:
```
Type: the kind of job run; analysis or production.
TaskType: A more detailed task type classification, based on the CMSSW config. Typically, analysis, DIGI, RECO, DIGI-RECO, GEN-SIM, or Cleanup.
```

The `Type` attribute (in the HTCondor job monitoring) comes from the `CMS_JobType` attribute that is set in WMAgent. As Federica said, it's a very generic activity that has a small number of well defined possibilities. The WMCore possible values are:
https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/WMBS/CreateWMBSBase.py#L503

while T0 adds a couple more:
https://github.com/dmwm/T0/blob/master/src/python/T0/WMBS/Oracle/Create.py#L841

So I'd suggest CRAB to use the same attribute `CMS_JobType`, with the generic values that make sense (anything else than analysis?)

Just to complement, `TaskType` from the job monitoring follows a logic/regex parsing the task name and trying to map it to a more specific task. It's know not to work for all cases though.

**Comment by stiegerb** [comment]
@amaltaro The `Type` in the job monitoring is just "analysis" or "production", and it's determined with a simple logic looking for analysis jobs. Everything else just has `Type` "production". See here:
https://github.com/dmwm/cms-htcondor-es/blob/master/src/htcondor_es/convert_to_json.py#L537
https://github.com/dmwm/cms-htcondor-es/blob/master/src/htcondor_es/convert_to_json.py#L551-L554

Then if the `Type` is "analysis", `CMS_JobType` is set to "Analysis", and to the classad value of `CMS_JobType` otherwise:
https://github.com/dmwm/cms-htcondor-es/blob/master/src/htcondor_es/convert_to_json.py#L591

**Comment by leggerf** [comment]
in the old dashboard, generic activities are not only Production and 
Analysis, but also: Integration, Test, SWinstallation, and unknown. 
Infact, I would have naively thought that HC jobs would be in Type: Testing.

WHat is the reason to have now only ANalysis and Production? Are the 
other categories 'obsolete'?

Il 24/10/2018 12:43, Alan Malta Rodrigues ha scritto:

**Comment by amaltaro** [comment]
I wasn't aware we had only those 2 values being pushed to MonIT.
@bbockelm is there any reason to completely bypass all the other job types defined in WMCore/T0/etc?

**Comment by belforte** [comment]
@amaltaro you misunderstood. I can explain. Sorry for making this a sort of bit thing, I did not want to take decision in isolation and push Benjaming myself to do. As I said what I really care about is to have the process defined. I think I know rather  well what all those variables are.

**Comment by belforte** [comment]
@vkuznet all classAds useful to monitoring should be defined there, CRAB+WMA no separate file, simply that file is needs being expanded and maintained (or moved elsewhere), we are discussion the process by which this should be done.
@leggerf old dashboards predates HTCondor and classAds, all things you see are still being fed via apMon to MonAlisa as UDP messages, than collected and stored at CERN by dashboards

**Comment by stiegerb** [comment]
Ok, I think we are mixing up two issues here: defining a meaningful type/activity field for analysis jobs and making the existing fields for production jobs more useful. The first is what this ticket is about. The second we should probably discuss somewhere else.

Relevant for this issue is that all type/activity fields we currently have are only filled with anything useful for production jobs. For analysis jobs they are all simply filled as "Analysis".

In particular, there is a `CRAB_JobType` field in the ads which is always set to "analysis". So we could start by renaming that field to `CMS_JobType`, and filling it with something meaningful.

For refining the type/activity of production jobs, shall we open a JIRA ticket?

**Comment by belforte** [comment]
proposal:

- [ ] @stiegerb stops overriding CMS_JobType for analysis
- [ ] @todor-ivanov changes CRAB to report CMS_JobType, set to `hctest` for HammerCloud and to `userAnalysis` otherwise
- [ ] @belforte looks into CRAB_JobType to make sure it is useless before we remove it
- [ ] about `data.Type = analysis/production` only in ES now, I do not see any need to change

for anything more, yes, would be useful to have a way to iterate on how to improve and expant your ReadMe, not usre JIRA is best, but whatever people like. This is a call for Valenting and Federica to make. I will fill in the info about CRAB-specific classAds

**Comment by amaltaro** [comment]
I;d make a correction:
> @todor-ivanov changes CRAB to report CMS_JobType, set to hctest for HammerCloud and to userAnalysis otherwise

keep using the `analysis` value instead of creating a new one. For HammerCloud, I don't remember what was the activity we have been using until now, but we better stick to exactly the same one (I think it was `hctest` indeed).

Then for Benjamin, we need to map CRAB jobs (CMS_JobType) to the kibana TaskType.

**Comment by belforte** [comment]
I thougt that want to overcame the euristic way to figure out TaskType with its known
problems in properly labelling production workflows e.g. as DIGIRECO and rather
set it in the submitter. No ?

**Comment by belforte** [comment]
why ? for historical compatibilyt ? I'd say.. fuck the history and do things properly from now on.
Otherwise we force CRAB to use CMS_JobType=analysis forever. If you
want an ad which only you can change the variable of, call it CMS_WMAjobType

**Comment by leggerf** [comment]
this will fuck up all historical queries though...can we reimport the 
old data and fix it?

Il 24/10/2018 16:57, Stefano Belforte ha scritto:

**Comment by belforte** [comment]
no,. simply we need different selections to plot data before/after the change.
But that is true even if we introduce a new variable.
Or maybe I am confused ?

**Comment by stiegerb** [comment]
@leggerf I don't think historical queries are a problem here. These fields were useless so far for analysis jobs. If someone really did query for `CMS_JobType = "Analysis"` then they could just as well switch to `Type = "Analysis"` now.

On the other hand, I'm not sure why "userAnalysis" is more proper than "Analysis".

**Comment by belforte** [comment]
because I'd like to have userMC as well instead of relaying on parsing inputdataset
an maybe also do userCustomScript.

**Comment by stiegerb** [comment]
Isn’t that rather a reason to keep it as “Analysis” for now until we have
those additional subcategories defined?

**Comment by belforte** [comment]
+1
I was worried about an undefine, but now that you showed how wrong I was... I fully agree. I will edit the tasl list in comment above https://github.com/dmwm/CRABServer/issues/5666#issuecomment-432678554

**Comment by leggerf** [comment]
my worry is: say somebody wants to get only analysis jobs. Query for 
analysis && NOT hc. This would retrieve from (let's say) today on really 
only analysis, but for the past analysis+HC jobs as this wasn't defined 
before. Applying two different queries based on time ranges is not 
really possible in graphana AFAIK


Il 24/10/2018 17:46, Benjamin Stieger ha scritto:

**Comment by belforte** [comment]
People who use the grafana dashboard to look in the past will need to be
aware that data are completed, validated, documented, only from some
point on. In general we can not make it possible to discriminate
old data based on fields which did not exist at the time.
Anyhow, as things stand now, data.Type=Analysis is idential to "CRAB"
and selects user and HC jobs as in the past. We simply introduce
a new distinction based on CMS_JobType

**Comment by leggerf** [comment]
yes, which then allows to make the distinction. Anyhow, since we're 
reviewing things, I propose to use the same classification in Type as in 
generic activities as in the old dashboard. These are 5-6 high level 
descriptions that anybody should be able to understand, and then in 
CMStype we store expert-only values (I let the experts decide if the 
list of values from the old dashboard still make sense, or if they need 
to be reviewed).

Il 24/10/2018 21:39, Stefano Belforte ha scritto:

**Comment by belforte** [comment]
I suggest to follow up in monitoring JIRA on the old/new taxonomy issue.

**Comment by belforte** [comment]
@todor-ivanov at the end , while waiting for final blessing by all, with 99% confidence we have converged on `CMS_JobType=hctest`
Anyhow I suggest that you test the code, but do not make a PR until the naming is finalized and documented (and signed in blood)

**Comment by todor-ivanov** [comment]
@stefano - ok I will change the name.

**Comment by belforte** [comment]
as discussed in mail:
```
CRAB_JobType=hctest/userAnalysis/userMC/userScript
i.e. CRAB_JobType will always be defined.
```

to beign with, currently CRAB_JobType classAd is defined and set to "analysis"
```
belforte@vocms0198/~> condor_q -con 'jobstatus==2&&jobuniverse==5' -af CRAB_JobType|sort|uniq -c
   5835 analysis
belforte@vocms0198/~> 
```
this needs to be set in DagmanCreator.py
https://github.com/dmwm/CRABServer/blob/f3f0acac35568c65cf67cbc576a89da9e6e07407/src/python/TaskWorker/Actions/DagmanCreator.py#L71

and the needed information is to be picked from the submission informaiton. I do not know exactly from where, I will look this up later or tomorrow

**Comment by amaltaro** [comment]
Wasn't it decided it would be called `CMS_JobType`, such that it is consistent with what central production defines.

**Comment by belforte** [comment]
I think @vkuznet things that what is defined here is so different from WMA JobType's that he wants a new name. I understand that there will be a document to review, and we can always change. For the time being let's simply have the correct code which sets the correct value.

**Comment by amaltaro** [comment]
Of course it's different, it's a analysis-like activity.
I still think keeping the same job classad name would be better, especially when we talk about a future where CRAB will be in WMCore somehow. Anyways, I think this topic has been exceptionally over discussed this week.

**Comment by belforte** [comment]
Bottom line of long discussion see:
https://its.cern.ch/jira/browse/CMSMONIT-28
https://its.cern.ch/jira/browse/CMSMONIT-57
https://its.cern.ch/jira/browse/CMSMONIT-59

- [ ] CRAB will detect HammerCloud submission when `activity` paramenter in submission configuration file is set to a string starting with `HC`
- [ ] CRAB will set the following two classAds:
- CMS_WMTool
set to:
     USER or HammerCloud based on the 'activity' parameter in crab configuration
- CMS_TaskType
set to:
     1 scriptExe or CMSRun according to crabConfig for user submission
     2 the activity field (e.g. HCtest or HCother) for HC submissions

**Comment by belforte** [comment]
@todor-ivanov you mentioned that you have a PR ready for this or at least part of it. Please put a pointer here so that I can review and possibly finalize and merge. Or let me know if I should go ahead and do this myself.

**Comment by todor-ivanov** [comment]
@belforte  Here [1] is the link to the commit  in my fork  there is only  the last step to fill in marked with a comment. And this was exactly my question at the crabdev meeting: 'How to find this piece exactly?'

[1]
https://github.com/todor-ivanov/CRABServer/commit/439f077bb2dd822f053ff95fedab458186cb26ea

**Comment by belforte** [comment]
thanks. Indeed some more code is needed to tell scriptExe vs cmsRun. I am importing your edits into my development version and will take care

**Comment by belforte** [comment]
so far so good, I ran a task through new code (which includex fix for #5813 ) and in the grid jobs the desired ads are set.
Need to verify scriptExe and all various use cases and to check with HC
```
belforte@vocms059/~> condor_q 576204.0 -l|grep CMS_
CMS_ALLOW_OVERFLOW = "True"
CMS_JobType = "analysis"
CMS_TaskType = "cmsRun"
CMS_WMTool = "USER"
belforte@vocms059/~> 
```
